File: src/main/go/aoc2019/day06.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
	"time"
)

type Orbit struct {
	inner, outer string
}

func Run06() {
	t := time.Now()

	orbits := read06()
	// fmt.Println(orbits[2].inner, orbits[2].outer)
	fmt.Println("read ", len(orbits), " orbits")

	fmt.Println("--- part 1")
	cnt := countOrbits(orbits, 0, "COM")
	fmt.Println(cnt)

	fmt.Println("--- part 2")
	tra := minTransfers(orbits, "YOU", "SAN") - 2
	fmt.Println(tra)

	fmt.Printf("=== end (%d ms)\n", (time.Now().Sub(t).Milliseconds()))
}

func countOrbits(orbits []Orbit, ci int, inner string) int {
	c := 0
	for _, o := range orbits {
		if o.inner == inner {
			c += ci + 1 + countOrbits(orbits, ci+1, o.outer)
		}
	}
	return c
}

func minTransfers(orbits []Orbit, start, target string) int {
	inf := 1000000
	distmap := make(map[string]int)
	distmap[start] = 0

	dst := func(x string) int {
		val, exists := distmap[x]
		if !exists {
			val = inf
			distmap[x] = val
		}
		return val
	}
	better := true
	for better {
		better = false
		for _, o := range orbits {
			di := dst(o.inner)
			do := dst(o.outer)
			if di > 1+do {
				better = true
				distmap[o.inner] = 1 + do
			}
			if do > 1+di {
				better = true
				distmap[o.outer] = 1 + di
			}
		}
	}

	return distmap[target]
}

func read06() []Orbit {
	fileBytes, _ := ioutil.ReadFile("input06.txt")
	lines := strings.Split(string(fileBytes), "\r\n")

	orb := make([]Orbit, 0, len(lines))
	for _, l := range lines {
		s := strings.Split(l, ")")
		orb = append(orb, Orbit{s[0], s[1]})
	}
	return orb
}

File: src/main/julia/day06.jl
data = [("COM", "B"),("B", "C"),("C", "D"),("D", "E"),("E", "F"),("B", "G"),("G", "H"),("D", "I"),("E", "J"),("J", "K"),("K", "L")] # ,("K", "YOU"),("I", "SAN")]

data = map(x->(x[1], x[2]), map(x->split(x,')'), readlines("input06.txt")))
# data = readlines("input06.txt") |> x->split.(x,')') |> x->tuple.(x[1], x[2])

outer(p) = map(x->x[2], filter(x->x[1] == p, data))
inner(p) = map(x->x[1], filter(x->x[2] == p, data))

function norbits(i, n)::Int
    o = outer(i)
    map(x -> n + 1 + norbits(x, n + 1), o) |> sum
end

@time norbits("COM", 0)

# part 2 --------------------------------

data = [("COM", "B"),("B", "C"),("C", "D"),("D", "E"),("E", "F"),("B", "G"),("G", "H"),("D", "I"),("E", "J"),("J", "K"),("K", "L") ,("K", "YOU"),("I", "SAN")]

function path(from, to, way)::Number
    if from==to 
        println(way)
        return 0
    end
    via = [inner(from)..., outer(from)...]
    via = filter(x-> x ∉ way, via)
    if length(via)==0
        return Inf
    end
    pathx(x) = 1 + path(x, to, [way..., x])
    paths = map(pathx, via)
    return minimum(paths)
end

function path(from, to) 
    from = inner(from)[1]
    to = inner(to)[1]
    path(from, to, [from])
end

path("B", "B")
path("G", "C")
path("C", "K")
path("X", "COM")
@time path("YOU", "SAN")

